openage/convert/opus/opusenc.pyx

```diff
@@ -0,0 +1,428 @@
+    op.packet = <unsigned char *> PyMem_Malloc(outbuf_sz)
+    if padbuf == NULL or op.packet == NULL:
+        free_buffers(&os, oe, padbuf, op.packet)
+        raise MemoryError("Could not allocate buffers.")
+    op.granulepos = <ogg.ogg_int64_t> inopt['pre_skip']
+
+    # Main encoding loop (one frame per iteration)
+    cdef opus.opus_int32 enc_bytes = 0  # encoded bytes in this iteration
+    cdef size_t ttl_samples = 0         # read samples
+    cdef size_t in_sz = len(inputdata)
+    cdef size_t in_pos = 0              # offset of next sample in inputdata
+    cdef size_t nb_samples              # no. of samples available this iteration
+    dbg("Starting encoding loop.")
+    while not op.e_o_s:
+        # fill buffer. Read directly from input_data, if possible.
+        nb_samples = min((in_sz - in_pos) // wframe_len, frame_sz)
+        inbuf = (<unsigned char *> inputdata) + ttl_samples * wframe_len
+        if nb_samples < frame_sz:
+            op.e_o_s = 1
+            memcpy(padbuf, inbuf, in_sz - in_pos)
+            memset(padbuf + in_sz - in_pos, 0, padbuf_sz - (in_sz - in_pos))
+            inbuf = padbuf
+
+        in_pos += nb_samples * wframe_len
+        ttl_samples += nb_samples
+
+
+        # convert.
+        enc_bytes = opus.opus_encode(oe, <opus.opus_int16 *> inbuf, frame_sz, op.packet, outbuf_sz)
+        # check how much was converted/ loop or EOF ?
+        if enc_bytes < 0:
+            free_buffers(&os, oe, padbuf, op.packet)
+            return "Encoding error in opus_encode(): {}".format(enc_bytes)
+
+        # append converted opuspacket.
+        op.bytes = enc_bytes
+        op.granulepos += nb_samples * (48000 / coding_rate)
+        if op.e_o_s:
+            pass  # TODO. set granulepos ? for resampling decoders.
+        op.packetno += 1
+
+        err = ogg.ogg_stream_packetin(&os, &op)
+        if err:
+            free_buffers(&os, oe, padbuf, op.packet)
+            return "ogg_stream_packetin() failed."
+
+        # Try to write page or force, if end of stream is reached.
+        while ogg.ogg_stream_pageout(&os, &og)\
+                or (op.e_o_s and ogg.ogg_stream_flush(&os, &og)):
+            outsize = write_ogg_page(outsize, outdata, &og)
+            if outsize == 0:
+                free_buffers(&os, oe, padbuf, op.packet)
+                raise MemoryError("Could not write opus data.")
+
+
+    free_buffers(&os, oe, padbuf, op.packet)
+
+    return outdata[:outsize]
+
+
+cdef int write_opus_header(ogg.ogg_stream_state *os, ogg.ogg_packet *op, inopt):
+    '''
+    Write the Opus header.
+    Returns non-zero on failure.
+    '''
+    try:
+        buf = bytearray(19)
+    except MemoryError:
+        return 1
+
+    buf[0:8] = b'OpusHead'
+    buf[8] = 1                # Opus encapsulating version.
+    buf[9] = inopt['channels']
+    buf[10:12] = inopt['pre_skip'].to_bytes(2, 'little')
+    buf[12:16] = inopt['input_rate'].to_bytes(4, 'little')
+    buf[16:18] = b'\x00\x00'  # Output gain
```

Comment:
I knew about `opusenc.html`, but (for some reason) didn't think that it would be useful.
Checked it now; it looks good.